File: backend/app/schemas/audit.py
"""
Audit-related Pydantic schemas.
"""
from typing import Any, Dict, Optional
from uuid import UUID

from pydantic import Field

from .base import BaseSchema, TimestampMixin, PaginatedResponse


class AuditLogResponse(TimestampMixin):
    """Schema for audit log response."""
    
    user_id: UUID = Field(description="User ID who performed the action")
    username: Optional[str] = Field(default=None, description="Username who performed the action")
    entity_type: str = Field(description="Type of entity that was modified")
    entity_id: UUID = Field(description="ID of the entity that was modified")
    operation: str = Field(description="Operation performed (CREATE, UPDATE, DELETE)")
    before_values: Optional[Dict[str, Any]] = Field(default=None, description="Values before the change")
    after_values: Optional[Dict[str, Any]] = Field(default=None, description="Values after the change")


class AuditLogListResponse(PaginatedResponse[AuditLogResponse]):
    """Schema for paginated audit log list response."""
    pass


class AuditLogFilters(BaseSchema):
    """Schema for audit log filtering parameters."""
    
    user_id: Optional[UUID] = Field(default=None, description="Filter by user ID")
    entity_type: Optional[str] = Field(default=None, description="Filter by entity type")
    entity_id: Optional[UUID] = Field(default=None, description="Filter by entity ID")
    operation: Optional[str] = Field(default=None, description="Filter by operation type")
    start_date: Optional[str] = Field(default=None, description="Filter by start date (ISO format)")
    end_date: Optional[str] = Field(default=None, description="Filter by end date (ISO format)")


class EntityChangesSummary(BaseSchema):
    """Schema for entity changes summary."""
    
    entity_type: str = Field(description="Type of entity")
    entity_id: UUID = Field(description="ID of the entity")
    total_changes: int = Field(description="Total number of changes")
    first_change: str = Field(description="Date of first change")
    last_change: str = Field(description="Date of last change")
    operations: Dict[str, int] = Field(description="Count of each operation type")


class UserActivitySummary(BaseSchema):
    """Schema for user activity summary."""
    
    user_id: UUID = Field(description="User ID")
    username: str = Field(description="Username")
    total_actions: int = Field(description="Total number of actions")
    first_action: str = Field(description="Date of first action")
    last_action: str = Field(description="Date of last action")
    operations: Dict[str, int] = Field(description="Count of each operation type")
    entities_modified: Dict[str, int] = Field(description="Count of entities modified by type")

File: frontend/src/api/audit.ts
import apiClient from './client'

export interface AuditLog {
  id: string
  user_id: string
  username?: string
  entity_type: string
  entity_id: string
  operation: string
  before_values?: Record<string, any>
  after_values?: Record<string, any>
  created_at: string
  updated_at: string
}

export interface AuditLogListResponse {
  items: AuditLog[]
  total: number
  page: number
  size: number
  pages: number
  has_next: boolean
  has_prev: boolean
}

export interface UserActivitySummary {
  user_id: string
  username: string
  total_actions: number
  first_action: string
  last_action: string
  operations: Record<string, number>
  entities_modified: Record<string, number>
}

export const auditApi = {
  listAuditLogs: async (params?: {
    skip?: number
    limit?: number
    user_id?: string
    entity_type?: string
    entity_id?: string
    operation?: string
    start_date?: string
    end_date?: string
  }): Promise<AuditLogListResponse> => {
    const response = await apiClient.get('/audit', { params })
    return response.data
  },

  getEntityHistory: async (entityType: string, entityId: string): Promise<AuditLog[]> => {
    const response = await apiClient.get(`/audit/entity/${entityType}/${entityId}`)
    return response.data
  },

  getUserActivity: async (userId: string, limit: number = 100): Promise<AuditLog[]> => {
    const response = await apiClient.get(`/audit/user/${userId}`, {
      params: { limit }
    })
    return response.data
  },

  getPermissionChanges: async (userId?: string, limit: number = 100): Promise<AuditLog[]> => {
    const response = await apiClient.get('/audit/permissions', {
      params: { user_id: userId, limit }
    })
    return response.data
  },

  getUserActivitySummary: async (userId: string): Promise<UserActivitySummary> => {
    const response = await apiClient.get(`/audit/user/${userId}/summary`)
    return response.data
  },

  getRecentChanges: async (limit: number = 50, entityType?: string): Promise<AuditLog[]> => {
    const response = await apiClient.get('/audit/recent', {
      params: { limit, entity_type: entityType }
    })
    return response.data
  },
}
